internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1033 @@ package controller
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/ptr"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	csiv1alpha1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	csiv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/ceph/ceph-csi-operator/internal/utils"
 )
 
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=operatorconfigs,verbs=get;list;watch
+
+type DriverType string
+
+const (
+	RbdDriverType    = "rbd"
+	CephFsDriverType = "cephfs"
+	NfsDriverType    = "nfs"
+)
+
+const ownerRefAnnotationKey = "csi.ceph.io/ownerref"
+
+// A regexp used to parse driver short name and driver type from the
+// driver's full name
+var nameRegExp, _ = regexp.Compile(fmt.Sprintf(
+	`^(.*)\.(%s|%s|%s)\.csi\.ceph\.com$`,
+	RbdDriverType,
+	CephFsDriverType,
+	NfsDriverType,
+))
+
 // DriverReconciler reconciles a Driver object
 type DriverReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+// A local reconcile object tied to a single reconcile iteration
+type driverReconcile struct {
+	DriverReconciler
+
+	ctx          context.Context
+	log          logr.Logger
+	driver       csiv1a1.Driver
+	driverPrefix string
+	driverType   DriverType
+	images       map[string]string
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DriverReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	// Define conditions for an OperatorConfig change that the require queuing of reconciliation
+	// request for drivers
+	driverDefaultsPredicate := builder.WithPredicates(
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				opConf, ok := e.Object.(*csiv1a1.OperatorConfig)
+				return ok && opConf.Spec.DriverSpecDefaults != nil
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				old, oldOk := e.ObjectOld.(*csiv1a1.OperatorConfig)
```

Comment:
What would you expect the behavior to be in that case?

Reply:
I expect not to fail on `old` being null.

Reply:
@leelavg I am failing to understand how this code will fail if either `old` or `new` is null (even before the new changes I introduced)

Reply:
oh, I said `old` rather than `oldOk`. My point is, there'll be no reconcile when `e.ObjectOld` could be `nil`.

However, as mentioned at the start, this is an edge case, can be resolved.

Reply:
Again, I feel I am missing something here: 
This is how the code looks like:
```
return !oldOk || !newOk || !reflect.DeepEqual(oldConf.Spec.DriverSpecDefaults, newConf.Spec.DriverSpecDefaults)
```
If either `oldOk` or `newOk` are nil we will return `true` which means a reconcile will happen.

Reply:
resolve, when I see it now, it seems I read the return statement wrongly.